webui/ocw/lib/EC2.py

```diff
@@ -5,22 +5,22 @@
 
 
 class EC2:
-    __instance = None
+    __instances = dict()
     __key = None
     __secret = None
     __ec2_resource = dict()
     __ec2_client = dict()
     __credentials = None
     __logging = None
 
-    def __new__(cls):
-        if EC2.__instance is None:
-            EC2.__instance = object.__new__(cls)
-            EC2.__instance.__credentials = EC2Credential()
-            EC2.__instance.__logging = logging.getLogger(__name__)
+    def __new__(cls, vault_namespace):
```

Comment:
something broken here from OOP perspective . Or  at least from Java-like OOP perspective . If we stating that class EC2 is keeping **collection** of all instances new method need to accept collection of namespaces  and return collection of instances. or we need another class e.g.  EC2_collection

Reply:
I think this is a classical case for a `Factory` with cache, but I was to lazy to implement it and used a `singleton` approach. Actually `singleton` makes also sense here, as we do wont this kind of object only once in the application.
But if you don't like it, I would make a proposal with a `Factory`. Maybe next PR, as this is quite big?

Reply:
this is my point you can not call this singleton ( again maybe just some weird python/perl singleton but not **that** https://en.wikipedia.org/wiki/Singleton_pattern singelton ) , because : 
1. singleton is not allowing to call `new` , you can get instance of object only via getInstance() 
2. singleton ( as it comes from its name  ) always having **single** instance here you having array of instances

Reply:
https://www.tutorialspoint.com/python_design_patterns/python_design_patterns_singleton.htm python version of SIngleton .

Reply:
Then lets call it decorator of a collection from singletons :)

There are some more ways to make singletons in python: https://python-3-patterns-idioms-test.readthedocs.io/en/latest/Singleton.html
And as you are not able to protect a constructor in python, you need othere ways as described in yours or mine link.
Not sure why I used a different one then in yours. Maybe because I thought it doesn't make sense to add a getInstance() call when I'm not able to hide the constructor at all.

Anyway, we need to say if we can keep it or not.

Reply:
> Then lets call it decorator of a collection from singletons :)

this discussion goes too deep into theoretical field :P 

> Anyway, we need to say if we can keep it or not.

totally agree . But let's conclude thread  around `sync_instance_db`  and "useless" namespace there . this would influence a lot in my opinion here ...